src/http/client.cr

```diff
@@ -625,23 +629,25 @@ class HTTP::Client
   end
 
   private def exec_internal(request, &block : Response -> T) : T forall T
-    exec_internal_single(request) do |response|
-      if response
-        return handle_response(response) { yield response }
-      end
-
-      # Server probably closed the connection, so retry once
-      close
-      request.body.try &.rewind
+    begin
       exec_internal_single(request) do |response|
         if response
-          return handle_response(response) do
-            yield response
-          end
+          return handle_response(response) { yield response }
         end
       end
+    rescue IO::Error
+      # retry
+    end
```

Comment:
This is a bit tricky: The `rescue` should not cover `yield`. Raising an exception in the block must not trigger a retry. Only exceptions raised in `send_request` should be rescued in order to try again.
I think for this it's probably best to move the call to `send_request` out of `exec_internal_single`. At this point, maybe the latter method could be removed entirely? Not sure.

Reply:
Good point, fixed, and added a spec for it